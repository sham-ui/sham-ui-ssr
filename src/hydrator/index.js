import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,

        // Pass default hooks service for proxy some methods (rehydrate & resolveID)
        new ServerSideHooks( DI.resolve( HOOKS_DI_KEY ), storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,

        // Don't pass default dom service, because ServerSideDOM full override it
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
