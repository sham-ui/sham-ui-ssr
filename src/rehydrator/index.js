import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
