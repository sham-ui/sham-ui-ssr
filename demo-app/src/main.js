import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

