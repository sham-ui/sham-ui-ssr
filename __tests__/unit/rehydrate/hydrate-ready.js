import { compileAsSFC, compile } from 'sham-ui-test-helpers';
import { ssrAndRehydrate } from '../../../src/testing';

it( 'defer hydrating with hydrateReady', async() => {
    expect.assertions( 2 );

    const meta = await ssrAndRehydrate(
        compileAsSFC( {
            Loader: compile`Loading...`
        } )`
            <template>
                {% if dataLoaded %}
                    Data loaded!
                {% else %}
                    <Loader/>
                {% endif %}
            </template>
            
            <script>
                export default Component( Template, function( options, didMount ) {
                    const dataLoaded = $();
                    
                    const state = options( {
                        [ dataLoaded ]: false
                    } );
                    
                    didMount( () => {
                        if ( !state[ dataLoaded ] ) {
                            this.hydrateReady = loadData().then(
                                loadDataSuccess    
                            );
                        }
                    } );
                    
                    function loadData() {
                        return new Promise( resolve => {
                            setTimeout( resolve, 100 );
                        } )
                    }
                    
                    const loadDataSuccess = () => state[ dataLoaded ] = true;
                } );
            </script>
        `
    );
    expect( meta.html ).toBe( ' Data loaded! <!--0-->' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );


it( 'defer hydrating with hydrateReady with element wrapper', async() => {
    expect.assertions( 2 );

    const meta = await ssrAndRehydrate(
        compileAsSFC( {
            Loader: compile`Loading...`
        } )`
            <template>
                <div>
                    <p>Page state</p>                
                    {% if dataLoaded %}
                        <p>Data loaded!</p>
                    {% else %}
                        <p>
                            <Loader/>
                        </p>
                    {% endif %}
                </div>
            </template>
            
            <script>
                export default Component( Template, function( options, didMount ) {
                    const dataLoaded = $();
                    
                    const state = options( {
                        [ dataLoaded ]: false
                    } );
                    
                    didMount( () => {
                        if ( !state[ dataLoaded ] ) {
                            this.hydrateReady = loadData().then(
                                loadDataSuccess    
                            );
                        }
                    } );
                    
                    function loadData() {
                        return new Promise( resolve => {
                            setTimeout( resolve, 100 );
                        } )
                    }
                    
                    const loadDataSuccess = () => state[ dataLoaded ] = true;
                } );
            </script>
        `
    );
    expect( meta.html ).toBe( '<div><p>Page state</p><p>Data loaded!</p><!--0--></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
