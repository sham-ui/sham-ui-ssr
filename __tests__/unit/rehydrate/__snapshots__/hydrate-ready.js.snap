// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defer hydrating with hydrateReady 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "dataLoaded": true,
  },
  "html": "
 Data loaded!
 <!--0-->
",
}
`;

exports[`defer hydrating with hydrateReady with element wrapper 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "dataLoaded": true,
  },
  "html": "
<div>
  <p>Page state</p>
  <p>Data loaded!</p>
  <!--0-->
</div>
",
}
`;
